### ISSUE TITLE
Many queued Dagruns created for DagRuns with a start_date in the past

### ISSUE THREAD

**Issue by SamWheating** [apache/airflow#18023]
### Apache Airflow version

2.1.3 (latest released)

### Operating System

Debian GNU/Linux 10 (buster)

### Deployment

Other 3rd-party Helm chart

### What happened

When a DAG is created with a start date far in the past and a relatively small schedule interval, the scheduler will create _a lot_ of queued DagRuns. 

This can potentially impact the time to move a DagRun from `queued` to `running`, as the `_start_queued_dagruns` function has to iterate through the entire set of `queued` DagRuns (looking at `max_dagruns_per_loop_to_schedule` at a time) before it re-examines a given DagRun. 

It appears that this is intended behaviour, but I'm wondering if we should be limiting the number of queued DagRuns per DAG in order to reduce strain on the scheduler. 

### What you expected to happen

In 2.1.2, the scheduler would only create `max_active_runs` DAGRuns for a single DAG, and then create additional runs as older runs were completed. 

### How to reproduce

This DAG will create ~2000 queued DagRuns when it is loaded:

```python
import time
from datetime import timedelta

from airflow import models
from airflow import utils
from airflow.operators import python_operator

dag = models.DAG(
    'data-infrastructure-examples.backfill-example-8',
    start_date=utils.timezone.utcnow() - timedelta(days=1),
    max_active_runs=1,
    dagrun_timeout=timedelta(minutes=10),
    schedule_interval="*/5 * * * *",
    catchup=True,
    concurrency=1
)

def sleep_five_minutes():
    print("sleeping 5m")
    time.sleep(300)
    print("Done")

child_task = python_operator.PythonOperator(
    task_id='sleep_5_minutes',
    dag=dag,
    python_callable=sleep_five_minutes)
```

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by collinmcnulty** [comment]
`max_active_dagruns` is checked when moving a dagrun from `queued` to `active`. My understanding is that centralizing the logic in `_start_queued_dagruns` was intentional to prevent there being multiple places that require logic for limiting the number of dagruns.

Are you seeing a performance impact? It would seem that while this increases the time between queued and running, changing this would only mean that same amount of time is spent in 'silently queued but invisible' and running.

**Comment by ephraimbuddy** [comment]
> This can potentially impact the time to move a DagRun from queued to running, as the _start_queued_dagruns function has to iterate through the entire set of queued DagRuns (looking at max_dagruns_per_loop_to_schedule at a time) before it re-examines a given DagRun.

It doesn't go through the entire dagruns. It goes through a subset controlled by `max_dagruns_per_loop_to_schedule`. And dagruns returned are ordered by `last_scheduling_decision`,  unless you are seeing a performance impact?

**Comment by ephraimbuddy** [comment]
It won't be a bad idea to add a limit to queued dagruns creation anyway

**Comment by SamWheating** [comment]
After further investigation, I think that most of the performance issues we saw were due to the issue now resolved by https://github.com/apache/airflow/pull/17945.

If needed, I can run some experiments and test the affects of having thousands of `queued` DagRuns on scheduler latency, but I think it's not as severe as I once thought. 

However, I still think that its a good idea to implement a limit for queued DagRuns, as creating thousands of queued runs in advance (specifically in the case of DAGs with a much earlier start date and `catchup=True`) can lead to some weird behaviour:

1) If the `end_date` is changed to an earlier date, there may already be queued runs after that date which have already been created.
2) If the schedule interval is changed, the change will not affect already created queued Dagruns. 
3) If there's a lifecycle policy on data in the DB, creating a DAGRun potentially weeks before it actually runs could cause the queued run to be dropped before it is ever run.
